File: tools/gba_header.ts
#!/usr/bin/env bun
import { readFileSync } from "fs";
import { dirname, resolve } from "path";
import { chunk } from "./export_asset.ts";
import { indexed_png } from "./import_asset.ts";
import { deflateSync } from "./zlib.ts";

export const GBA_HEADER_ADDRESS = 0x08000000;
export const GBA_HEADER_SIZE = 0xc0;
export const GBA_LOGO_ADDRESS = GBA_HEADER_ADDRESS + 0x04;
export const GBA_LOGO_SIZE = 0x9c;
export const GBA_FIXED_ADDRESS = GBA_HEADER_ADDRESS + 0xb2;
export const GBA_FIXED_SIZE = 0x0a;
export const GBA_RESERVED_END_ADDRESS = GBA_HEADER_ADDRESS + 0xbe;
export const GBA_RESERVED_END_SIZE = 0x02;
export const GBA_LOGO_WIDTH = 104;
export const GBA_LOGO_HEIGHT = 16;

const ROOT = dirname(dirname(Bun.fileURLToPath(import.meta.url)));
const LOGO_SHA256 = "08a0153cfd6b0ea54b938f7d209933fa849da0d56f5a34c481060c9ff2fad818";
const CODEWORDS = [
  "1", "0110", "01010", "0100", "00010", "011110", "010110", "000110",
  "00110", "011111", "010111", "000111", "0010", "01110", "00111", "0000",
] as const;

type JsonObject = Record<string, unknown>;

export interface GbaHeaderSource {
  format: 1;
  kind: "gba-cartridge-header-standard-fields";
  address: "0x08000000";
  standard: {
    logo: {
      codec: "gba-bios-huffman-logo";
      source: string;
      width: 104;
      height: 16;
      bpp: 1;
    };
    fixed_value: "0x96";
    unit_code: "0x00";
    device_type: "0x00";
    reserved_zero_ranges: Array<{ address: string; size: number }>;
  };
  unresolved_fields: Array<{ name: string; address: string; size: number }>;
}

const EXPECTED_RESERVED = [
  { address: "0x080000b5", size: 7 },
  { address: "0x080000be", size: 2 },
];

const EXPECTED_UNRESOLVED = [
  { name: "entry_branch", address: "0x08000000", size: 4 },
  { name: "title", address: "0x080000a0", size: 12 },
  { name: "game_code", address: "0x080000ac", size: 4 },
  { name: "maker_code", address: "0x080000b0", size: 2 },
  { name: "software_version", address: "0x080000bc", size: 1 },
  { name: "complement_checksum", address: "0x080000bd", size: 1 },
];

function object(value: unknown, label: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  return value as JsonObject;
}

function exactKeys(value: JsonObject, expected: string[], label: string): void {
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  if (actual.length !== wanted.length || actual.some((key, index) => key !== wanted[index])) {
    throw new Error(`${label} fields differ`);
  }
}

function parseRecord(value: unknown, label: string, withName: boolean): { name?: string; address: string; size: number } {
  const record = object(value, label);
  exactKeys(record, withName ? ["name", "address", "size"] : ["address", "size"], label);
  if (withName && typeof record.name !== "string") throw new Error(`${label} name differs`);
  if (typeof record.address !== "string" || !/^0x[0-9a-f]{8}$/.test(record.address)) {
    throw new Error(`${label} address is not canonical`);
  }
  if (!Number.isSafeInteger(record.size) || Number(record.size) <= 0) throw new Error(`${label} size differs`);
  return { ...(withName ? { name: String(record.name) } : {}), address: record.address, size: Number(record.size) };
}

function sameRecords(
  actual: Array<{ name?: string; address: string; size: number }>,
  expected: Array<{ name?: string; address: string; size: number }>,
): boolean {
  return actual.length === expected.length && actual.every((record, index) =>
    record.name === expected[index].name && record.address === expected[index].address && record.size === expected[index].size);
}

export function parseGbaHeaderSource(value: unknown): GbaHeaderSource {
  const source = object(value, "GBA header source");
  exactKeys(source, ["format", "kind", "address", "standard", "unresolved_fields"], "GBA header source");
  if (source.format !== 1 || source.kind !== "gba-cartridge-header-standard-fields" ||
      source.address !== "0x08000000") throw new Error("unsupported GBA header source");
  const standard = object(source.standard, "GBA header standard fields");
  exactKeys(standard, ["logo", "fixed_value", "unit_code", "device_type", "reserved_zero_ranges"],
    "GBA header standard fields");
  const logo = object(standard.logo, "GBA header logo");
  exactKeys(logo, ["codec", "source", "width", "height", "bpp"], "GBA header logo");
  if (logo.codec !== "gba-bios-huffman-logo" || typeof logo.source !== "string" || logo.source.length === 0 ||
      logo.width !== GBA_LOGO_WIDTH || logo.height !== GBA_LOGO_HEIGHT || logo.bpp !== 1 ||
      standard.fixed_value !== "0x96" || standard.unit_code !== "0x00" || standard.device_type !== "0x00") {
    throw new Error("GBA header standard values differ");
  }
  if (!Array.isArray(standard.reserved_zero_ranges)) throw new Error("GBA header reserved ranges differ");
  const reserved = standard.reserved_zero_ranges.map((record, index) =>
    parseRecord(record, `GBA header reserved range ${index}`, false));
  if (!sameRecords(reserved, EXPECTED_RESERVED)) throw new Error("GBA header reserved ranges differ");
  if (!Array.isArray(source.unresolved_fields)) throw new Error("GBA header unresolved fields differ");
  const unresolved = source.unresolved_fields.map((record, index) =>
    parseRecord(record, `GBA header unresolved field ${index}`, true));
  if (!sameRecords(unresolved, EXPECTED_UNRESOLVED)) throw new Error("GBA header unresolved fields differ");
  return value as GbaHeaderSource;
}

function pngHeader(width: number, height: number): Buffer {
  const header = Buffer.alloc(13);
  header.writeUInt32BE(width, 0);
  header.writeUInt32BE(height, 4);
  header.set([1, 3, 0, 0, 0], 8);
  return header;
}

export function logoPng(pixels: Uint8Array): Buffer {
  if (pixels.length !== GBA_LOGO_WIDTH * GBA_LOGO_HEIGHT ||
      pixels.some((pixel) => pixel !== 0 && pixel !== 1)) throw new Error("GBA logo pixels must be 104x16 1bpp");
  const rows = Buffer.alloc(GBA_LOGO_HEIGHT * (GBA_LOGO_WIDTH / 8 + 1));
  let cursor = 0;
  for (let y = 0; y < GBA_LOGO_HEIGHT; y++) {
    rows[cursor++] = 0;
    for (let x = 0; x < GBA_LOGO_WIDTH; x += 8) {
      let value = 0;
      for (let bit = 0; bit < 8; bit++) value |= pixels[y * GBA_LOGO_WIDTH + x + bit] << (7 - bit);
      rows[cursor++] = value;
    }
  }
  return Buffer.concat([
    Buffer.from("89504e470d0a1a0a", "hex"),
    chunk(Buffer.from("IHDR"), pngHeader(GBA_LOGO_WIDTH, GBA_LOGO_HEIGHT)),
    chunk(Buffer.from("PLTE"), Buffer.from("ffffff000000", "hex")),
    chunk(Buffer.from("IDAT"), deflateSync(rows, { level: 9 })),
    chunk(Buffer.from("IEND"), Buffer.alloc(0)),
  ]);
}

function tiledLogoBits(pixels: readonly number[]): Buffer {
  const decoded = Buffer.alloc(GBA_LOGO_WIDTH * GBA_LOGO_HEIGHT / 8);
  const tilesWide = GBA_LOGO_WIDTH / 8;
  for (let y = 0; y < GBA_LOGO_HEIGHT; y++) {
    for (let x = 0; x < GBA_LOGO_WIDTH; x++) {
      const tile = Math.floor(y / 8) * tilesWide + Math.floor(x / 8);
      const bit = tile * 64 + (y & 7) * 8 + (x & 7);
      decoded[bit >> 3] |= pixels[y * GBA_LOGO_WIDTH + x] << (bit & 7);
    }
  }
  return decoded;
}

function additionDeltas(decoded: Buffer): Buffer {
  if (decoded.length % 2 !== 0) throw new Error("GBA logo bitmap must contain whole halfwords");
  const output = Buffer.alloc(decoded.length + 4);
  output.writeUInt32LE(decoded.length << 8 | 0x82, 0);
  let previous = 0;
  for (let offset = 0; offset < decoded.length; offset += 2) {
    const current = decoded.readUInt16LE(offset);
    output.writeUInt16LE((current - previous) & 0xffff, offset + 4);
    previous = current;
  }
  return output;
}

function huffmanLogo(data: Uint8Array): Buffer {
  const output = Buffer.alloc(GBA_LOGO_SIZE);
  let position = 0;
  for (const value of data) {
    for (const nibble of [value & 15, value >>> 4]) {
      for (const character of CODEWORDS[nibble]) {
        if (position >= 0x4ce) throw new Error("GBA logo exceeds its fixed Huffman field");
        if (character === "1") {
          const wordOffset = Math.floor(position / 32) * 4;
          const shift = 31 - position % 32;
          const current = output.readUInt32LE(wordOffset);
          output.writeUInt32LE((current | (1 << shift)) >>> 0, wordOffset);
        }
        position++;
      }
    }
  }
  output[0x98] = 0x21;
  output[0x99] = 0xd4;
  return output;
}

export function encodeGbaLogo(image: Uint8Array): Buffer {
  const encodedImage = Buffer.from(image);
  const [width, height, pixels, palette] = indexed_png(encodedImage);
  if (width !== GBA_LOGO_WIDTH || height !== GBA_LOGO_HEIGHT ||
      !Bun.deepEquals(palette, [[255, 255, 255], [0, 0, 0]], true) ||
      !encodedImage.equals(logoPng(Uint8Array.from(pixels)))) {
    throw new Error("GBA logo source must be the canonical 104x16 monochrome PNG");
  }
  const output = huffmanLogo(additionDeltas(tiledLogoBits(pixels)));
  const hash = new Bun.CryptoHasher("sha256").update(output).digest("hex");
  if (hash !== LOGO_SHA256) throw new Error("GBA logo source does not encode the standard firmware logo");
  return output;
}

export function buildGbaHeaderComponent(
  value: unknown,
  logoImage: Uint8Array,
  address: number,
  size: number,
): Buffer {
  parseGbaHeaderSource(value);
  if (address === GBA_LOGO_ADDRESS && size === GBA_LOGO_SIZE) return encodeGbaLogo(logoImage);
  if (address === GBA_FIXED_ADDRESS && size === GBA_FIXED_SIZE) {
    return Buffer.from([0x96, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]);
  }
  if (address === GBA_RESERVED_END_ADDRESS && size === GBA_RESERVED_END_SIZE) return Buffer.alloc(2);
  throw new Error("GBA header component is not independently determined");
}

export function encodeArmBranch(address: number, target: number): Buffer {
  if (!Number.isSafeInteger(address) || !Number.isSafeInteger(target) || address < 0 || target < 0 ||
      address > 0xfffffffc || target > 0xfffffffc || (address & 3) !== 0 || (target & 3) !== 0) {
    throw new Error("ARM branch addresses must be aligned u32 values");
  }
  const words = (target - (address + 8)) / 4;
  if (!Number.isInteger(words) || words < -0x800000 || words > 0x7fffff) {
    throw new Error("ARM branch target is out of range");
  }
  const output = Buffer.alloc(4);
  output.writeUInt32LE((0xea000000 | (words & 0x00ffffff)) >>> 0);
  return output;
}

export function gbaComplementChecksum(header: Uint8Array): number {
  if (header.length < 0xbd) throw new Error("GBA header is too short for its complement checksum");
  let checksum = -0x19;
  for (let offset = 0xa0; offset <= 0xbc; offset++) checksum -= header[offset];
  return checksum & 0xff;
}

function selfTest(): void {
  const sourcePath = resolve(ROOT, "assets/data/gba_header.json");
  const source = parseGbaHeaderSource(JSON.parse(readFileSync(sourcePath, "utf8")));
  const logoPath = resolve(ROOT, source.standard.logo.source);
  const logo = buildGbaHeaderComponent(source, readFileSync(logoPath), GBA_LOGO_ADDRESS, GBA_LOGO_SIZE);
  if (logo.length !== GBA_LOGO_SIZE) throw new Error("GBA logo self-test failed");
  const fixed = buildGbaHeaderComponent(source, Buffer.alloc(0), GBA_FIXED_ADDRESS, GBA_FIXED_SIZE);
  if (fixed.toString("hex") !== "96000000000000000000") throw new Error("GBA fixed-field self-test failed");
  if (buildGbaHeaderComponent(source, Buffer.alloc(0), GBA_RESERVED_END_ADDRESS, GBA_RESERVED_END_SIZE).some(Boolean)) {
    throw new Error("GBA reserved-field self-test failed");
  }
  if (encodeArmBranch(0x08000000, 0x080000c0).toString("hex") !== "2e0000ea") {
    throw new Error("ARM branch self-test failed");
  }
  if (gbaComplementChecksum(Buffer.alloc(GBA_HEADER_SIZE)) !== 0xe7) throw new Error("GBA checksum self-test failed");
  const reject = (callback: () => unknown): void => {
    try { callback(); } catch { return; }
    throw new Error("invalid GBA header source was accepted");
  };
  reject(() => buildGbaHeaderComponent(source, Buffer.alloc(0), GBA_HEADER_ADDRESS, 4));
  reject(() => parseGbaHeaderSource({ ...source, standard: { ...source.standard, fixed_value: "0x00" } }));
  reject(() => encodeArmBranch(0x08000002, 0x080000c0));
}

if (import.meta.main) {
  if (!process.argv.includes("--self-test")) throw new Error("usage: gba_header.ts --self-test");
  selfTest();
  console.log("gba header tests passed");
}

File: tools/alchemy/src/build_assets/gba_header.rs
//! GBA cartridge-header codec. Like pret's `rom_header.s`, the source reserves
//! the 156-byte cartridge logo instead of describing it: the logo is a private
//! input extracted from the verified ROM and placed by its own region.

use std::fs;
use std::path::Path;

use crate::compiler::canonical_json::is_canonical_json_text;
use serde_json::{Map, Value};

const GBA_HEADER_ADDRESS: u32 = 0x0800_0000;
const GBA_HEADER_SIZE: usize = 0xc0;
const GBA_LOGO_OFFSET: usize = 0x04;
const GBA_LOGO_SIZE: usize = 0x9c;
const GBA_FIXED_ADDRESS: u32 = GBA_HEADER_ADDRESS + 0xb2;
const GBA_FIXED_SIZE: usize = 0x0a;
const GBA_RESERVED_END_ADDRESS: u32 = GBA_HEADER_ADDRESS + 0xbe;
const GBA_RESERVED_END_SIZE: usize = 0x02;

const EXPECTED_RESERVED: [(&str, usize); 2] = [("0x080000b5", 7), ("0x080000be", 2)];
const EXPECTED_UNRESOLVED: [(&str, &str, usize); 6] = [
    ("entry_branch", "0x08000000", 4),
    ("title", "0x080000a0", 12),
    ("game_code", "0x080000ac", 4),
    ("maker_code", "0x080000b0", 2),
    ("software_version", "0x080000bc", 1),
    ("complement_checksum", "0x080000bd", 1),
];

#[derive(Clone, Debug)]
struct Edition {
    target: u32,
    title: String,
    game_code: String,
    maker_code: String,
    software_version: u8,
}

fn object<'a>(value: &'a Value, label: &str) -> Result<&'a Map<String, Value>, String> {
    value
        .as_object()
        .ok_or_else(|| format!("{label} must be an object"))
}

fn exact_keys(value: &Map<String, Value>, expected: &[&str], label: &str) -> Result<(), String> {
    let mut actual: Vec<&str> = value.keys().map(String::as_str).collect();
    actual.sort_unstable();
    let mut wanted = expected.to_vec();
    wanted.sort_unstable();
    if actual != wanted {
        return Err(format!("{label} fields differ"));
    }
    Ok(())
}

fn number_equals(value: Option<&Value>, expected: u64) -> bool {
    let Some(value) = value else { return false };
    if let Some(actual) = value.as_u64() {
        return actual == expected;
    }
    value
        .as_f64()
        .is_some_and(|actual| actual.is_finite() && actual == expected as f64)
}

fn safe_positive_integer(value: Option<&Value>) -> Option<usize> {
    let number = value?.as_f64()?;
    if !number.is_finite()
        || number.fract() != 0.0
        || number <= 0.0
        || number > 9_007_199_254_740_991.0
    {
        return None;
    }
    Some(number as usize)
}

fn safe_integer(value: Option<&Value>) -> Option<i64> {
    let number = value?.as_f64()?;
    if !number.is_finite() || number.fract() != 0.0 || number.abs() > 9_007_199_254_740_991.0 {
        return None;
    }
    Some(number as i64)
}

fn canonical_address(value: Option<&Value>, label: &str) -> Result<(String, u32), String> {
    let text = value
        .and_then(Value::as_str)
        .ok_or_else(|| format!("{label} is not a canonical address"))?;
    if text.len() != 10
        || !text.starts_with("0x")
        || !text[2..]
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
    {
        return Err(format!("{label} is not a canonical address"));
    }
    let address = u32::from_str_radix(&text[2..], 16)
        .map_err(|_| format!("{label} is not a canonical address"))?;
    Ok((text.to_string(), address))
}

fn parse_record(
    value: &Value,
    label: &str,
    with_name: bool,
) -> Result<(Option<String>, String, usize), String> {
    let record = object(value, label)?;
    let keys = if with_name {
        &["name", "address", "size"][..]
    } else {
        &["address", "size"][..]
    };
    exact_keys(record, keys, label)?;
    let name = if with_name {
        Some(
            record
                .get("name")
                .and_then(Value::as_str)
                .ok_or_else(|| format!("{label} name differs"))?
                .to_string(),
        )
    } else {
        None
    };
    let (address, _) = canonical_address(record.get("address"), &format!("{label} address"))?;
    let size = safe_positive_integer(record.get("size"));
    if size.is_none() {
        return Err(format!("{label} size differs"));
    }
    Ok((name, address, size.unwrap()))
}

fn parse_ascii(value: Option<&Value>, length: usize, label: &str) -> Result<String, String> {
    let text = value
        .and_then(Value::as_str)
        .ok_or_else(|| format!("{label} differs"))?;
    if text.len() != length
        || !text
            .bytes()
            .all(|byte| byte.is_ascii_uppercase() || byte.is_ascii_digit())
    {
        return Err(format!("{label} differs"));
    }
    Ok(text.to_string())
}

fn title_bytes(title: &str) -> Result<[u8; 12], String> {
    let bytes = title.as_bytes();
    if bytes.len() > 12 || bytes.iter().any(|byte| !(*byte >= 0x20 && *byte <= 0x7e)) {
        return Err("GBA title must contain at most twelve printable ASCII bytes".to_string());
    }
    let mut output = [0u8; 12];
    output[..bytes.len()].copy_from_slice(bytes);
    Ok(output)
}

fn parse_edition(value: &Value) -> Result<Edition, String> {
    let edition = object(value, "GBA header edition")?;
    exact_keys(
        edition,
        &[
            "entry_branch",
            "title",
            "game_code",
            "maker_code",
            "software_version",
            "complement_checksum",
        ],
        "GBA header edition",
    )?;
    let entry = object(
        edition
            .get("entry_branch")
            .ok_or("GBA entry branch must be an object")?,
        "GBA entry branch",
    )?;
    exact_keys(
        entry,
        &["instruction_set", "operation", "target"],
        "GBA entry branch",
    )?;
    if entry.get("instruction_set").and_then(Value::as_str) != Some("arm")
        || entry.get("operation").and_then(Value::as_str) != Some("b")
    {
        return Err("GBA entry branch differs".to_string());
    }
    let (_, target) = canonical_address(entry.get("target"), "GBA entry target")?;
    encode_arm_branch(GBA_HEADER_ADDRESS, target)?;

    let title = object(
        edition.get("title").ok_or("GBA title must be an object")?,
        "GBA title",
    )?;
    exact_keys(title, &["text", "padding", "field_bytes"], "GBA title")?;
    let title_text = title
        .get("text")
        .and_then(Value::as_str)
        .ok_or("GBA title encoding differs")?;
    if title.get("padding").and_then(Value::as_str) != Some("nul")
        || !number_equals(title.get("field_bytes"), 12)
    {
        return Err("GBA title encoding differs".to_string());
    }
    title_bytes(title_text)?;
    let game_code = parse_ascii(edition.get("game_code"), 4, "GBA game code")?;
    let maker_code = parse_ascii(edition.get("maker_code"), 2, "GBA maker code")?;
    let version = safe_integer(edition.get("software_version"));
    if !version.is_some_and(|value| (0..=0xff).contains(&value)) {
        return Err("GBA software version differs".to_string());
    }
    if edition.get("complement_checksum").and_then(Value::as_str) != Some("derived") {
        return Err("GBA checksum policy differs".to_string());
    }
    Ok(Edition {
        target,
        title: title_text.to_string(),
        game_code,
        maker_code,
        software_version: version.unwrap() as u8,
    })
}

fn parse_gba_header_source(value: &Value) -> Result<(), String> {
    let source = object(value, "GBA header source")?;
    exact_keys(
        source,
        &[
            "format",
            "kind",
            "address",
            "standard",
            "edition",
            "unresolved_fields",
        ],
        "GBA header source",
    )?;
    if !number_equals(source.get("format"), 4)
        || source.get("kind").and_then(Value::as_str)
            != Some("gba-cartridge-header-standard-fields")
        || source.get("address").and_then(Value::as_str) != Some("0x08000000")
    {
        return Err("unsupported GBA header source".to_string());
    }
    let standard = object(
        source
            .get("standard")
            .ok_or("GBA header standard fields must be an object")?,
        "GBA header standard fields",
    )?;
    exact_keys(
        standard,
        &[
            "fixed_value",
            "unit_code",
            "device_type",
            "reserved_zero_ranges",
        ],
        "GBA header standard fields",
    )?;
    if standard.get("fixed_value").and_then(Value::as_str) != Some("0x96")
        || standard.get("unit_code").and_then(Value::as_str) != Some("0x00")
        || standard.get("device_type").and_then(Value::as_str) != Some("0x00")
    {
        return Err("GBA header standard values differ".to_string());
    }
    let reserved = source_array(
        standard.get("reserved_zero_ranges"),
        "GBA header reserved ranges differ",
    )?;
    let parsed_reserved: Vec<(Option<String>, String, usize)> = reserved
        .iter()
        .enumerate()
        .map(|(index, value)| {
            parse_record(value, &format!("GBA header reserved range {index}"), false)
        })
        .collect::<Result<_, _>>()?;
    if parsed_reserved.len() != EXPECTED_RESERVED.len()
        || parsed_reserved.iter().zip(EXPECTED_RESERVED).any(
            |((_, address, size), (expected_address, expected_size))| {
                address != expected_address || *size != expected_size
            },
        )
    {
        return Err("GBA header reserved ranges differ".to_string());
    }

    let unresolved = source_array(
        source.get("unresolved_fields"),
        "GBA header unresolved fields differ",
    )?;
    let parsed_unresolved: Vec<(Option<String>, String, usize)> = unresolved
        .iter()
        .enumerate()
        .map(|(index, value)| {
            parse_record(value, &format!("GBA header unresolved field {index}"), true)
        })
        .collect::<Result<_, _>>()?;
    match source.get("edition") {
        Some(Value::Null) => {
            if parsed_unresolved.len() != EXPECTED_UNRESOLVED.len()
                || parsed_unresolved.iter().zip(EXPECTED_UNRESOLVED).any(
                    |((name, address, size), (expected_name, expected_address, expected_size))| {
                        name.as_deref() != Some(expected_name)
                            || address != expected_address
                            || *size != expected_size
                    },
                )
            {
                return Err("GBA header unresolved fields differ".to_string());
            }
        }
        Some(edition) => {
            parse_edition(edition)?;
            if !unresolved.is_empty() {
                return Err("resolved GBA header still names unresolved fields".to_string());
            }
        }
        None => return Err("GBA header edition must be an object".to_string()),
    }
    Ok(())
}

fn source_array<'a>(value: Option<&'a Value>, message: &str) -> Result<&'a Vec<Value>, String> {
    value
        .and_then(Value::as_array)
        .ok_or_else(|| message.to_string())
}

pub(super) fn read_gba_header_source(path: &Path) -> Result<Value, String> {
    let text = fs::read_to_string(path).map_err(|error| error.to_string())?;
    let value: Value = serde_json::from_str(&text).map_err(|error| error.to_string())?;
    if !is_canonical_json_text(&text, &value) {
        return Err(format!("{}: source is not canonical JSON", path.display()));
    }
    parse_gba_header_source(&value)?;
    Ok(value)
}

fn encode_arm_branch(address: u32, target: u32) -> Result<Vec<u8>, String> {
    if address > 0xffff_fffc
        || target > 0xffff_fffc
        || !address.is_multiple_of(4)
        || !target.is_multiple_of(4)
    {
        return Err("ARM branch addresses must be aligned u32 values".to_string());
    }
    let words = (target as i64 - (address as i64 + 8)) / 4;
    if !(-0x80_0000..=0x7f_ffff).contains(&words) {
        return Err("ARM branch target is out of range".to_string());
    }
    let instruction = 0xea00_0000u32 | (words as u32 & 0x00ff_ffff);
    Ok(instruction.to_le_bytes().to_vec())
}

fn gba_complement_checksum(header: &[u8]) -> Result<u8, String> {
    if header.len() < 0xbd {
        return Err("GBA header is too short for its complement checksum".to_string());
    }
    let mut checksum: i64 = -0x19;
    for byte in &header[0xa0..=0xbc] {
        checksum -= i64::from(*byte);
    }
    Ok(checksum.rem_euclid(256) as u8)
}

fn edition(source: &Value) -> Result<Edition, String> {
    let source_object = object(source, "GBA header source")?;
    match source_object.get("edition") {
        Some(Value::Null) => Err("GBA header edition is unresolved".to_string()),
        Some(value) => parse_edition(value),
        None => Err("GBA header edition is unresolved".to_string()),
    }
}

/// The header with its logo field reserved as zero bytes; the checksum covers
/// only the fields after the logo.
fn build_gba_header(source: &Value) -> Result<Vec<u8>, String> {
    parse_gba_header_source(source)?;
    let edition = edition(source)?;
    let mut output = vec![0u8; GBA_HEADER_SIZE];
    output[0..4].copy_from_slice(&encode_arm_branch(GBA_HEADER_ADDRESS, edition.target)?);
    output[0xa0..0xac].copy_from_slice(&title_bytes(&edition.title)?);
    output[0xac..0xb0].copy_from_slice(edition.game_code.as_bytes());
    output[0xb0..0xb2].copy_from_slice(edition.maker_code.as_bytes());
    output[0xb2] = 0x96;
    output[0xbc] = edition.software_version;
    output[0xbd] = gba_complement_checksum(&output)?;
    Ok(output)
}

pub(super) fn build_gba_header_component(
    source: &Value,
    address: u32,
    size: usize,
) -> Result<Vec<u8>, String> {
    parse_gba_header_source(source)?;
    let logo = GBA_HEADER_ADDRESS as u64 + GBA_LOGO_OFFSET as u64;
    if (address as u64) < logo + GBA_LOGO_SIZE as u64 && logo < address as u64 + size as u64 {
        return Err("the cartridge logo is a registered private input, not a header field".into());
    }
    if address == GBA_FIXED_ADDRESS && size == GBA_FIXED_SIZE {
        return Ok(vec![0x96, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
    }
    if address == GBA_RESERVED_END_ADDRESS && size == GBA_RESERVED_END_SIZE {
        return Ok(vec![0, 0]);
    }
    if object(source, "GBA header source")?
        .get("edition")
        .is_some_and(|value| !value.is_null())
        && size > 0
        && address >= GBA_HEADER_ADDRESS
        && (address as u64 + size as u64) <= GBA_HEADER_ADDRESS as u64 + GBA_HEADER_SIZE as u64
    {
        let header = build_gba_header(source)?;
        let offset = (address - GBA_HEADER_ADDRESS) as usize;
        return Ok(header[offset..offset + size].to_vec());
    }
    Err("GBA header component is not independently determined".to_string())
}

#[test]
fn header_fields_build_around_the_reserved_logo() {
    let mut source = serde_json::json!({
        "format": 4, "kind": "gba-cartridge-header-standard-fields", "address": "0x08000000",
        "standard": {"fixed_value": "0x96", "unit_code": "0x00", "device_type": "0x00",
            "reserved_zero_ranges": [{"address": "0x080000b5", "size": 7}, {"address": "0x080000be", "size": 2}]},
        "edition": {"entry_branch": {"instruction_set": "arm", "operation": "b", "target": "0x080000c0"},
            "title": {"text": "TEST", "padding": "nul", "field_bytes": 12},
            "game_code": "ATST", "maker_code": "01", "software_version": 2, "complement_checksum": "derived"},
        "unresolved_fields": []
    });
    let branch = build_gba_header_component(&source, 0x0800_0000, 4).unwrap();
    assert_eq!(branch, 0xea00_002eu32.to_le_bytes());
    let fields = build_gba_header_component(&source, 0x0800_00a0, 0x20).unwrap();
    assert_eq!(&fields[..6], b"TEST\0\0");
    assert_eq!(&fields[0x0c..0x12], b"ATST01");
    assert_eq!((fields[0x12], fields[0x1c]), (0x96, 2));
    let sum = fields[..0x1e]
        .iter()
        .fold(0x19u8, |sum, byte| sum.wrapping_add(*byte));
    assert_eq!(sum, 0, "complement checksum");
    // No range reaching into the logo field builds, and no source may carry it.
    for (address, size) in [(0x0800_0000, 5), (0x0800_0004, 0x9c), (0x0800_009f, 2)] {
        assert!(build_gba_header_component(&source, address, size).is_err());
    }
    source["standard"]["logo"] = serde_json::json!({"source": "LOGO.BIN"});
    assert!(build_gba_header_component(&source, 0x0800_0000, 4).is_err());
    source["standard"].as_object_mut().unwrap().remove("logo");
    source["format"] = Value::from(3);
    assert!(build_gba_header_component(&source, 0x0800_0000, 4).is_err());
}
